cockatrice/src/client/ui/layouts/flow_layout.cpp

```diff
@@ -0,0 +1,331 @@
+    }
+
+    // Layout the final row if there are remaining items.
+    layoutSingleRow(rowItems, originX, currentYPosition);
+
+    // Return the total y-coordinate at the bottom of the last row.
+    return currentYPosition + rowHeight;
+}
+
+/**
+ * @brief Helper function for arranging a single row of items within specified bounds.
+ * @param rowItems Items to be arranged in the row.
+ * @param x The x-coordinate for starting the row.
+ * @param y The y-coordinate for starting the row.
+ */
+void FlowLayout::layoutSingleRow(const QVector<QLayoutItem *> &rowItems, int x, const int y)
+{
+    // Iterate through each item in the row and position it.
+    for (QLayoutItem *item : rowItems) {
+        if (item == nullptr || item->isEmpty()) {
+            continue;
+        }
+        QSize itemMaxSize = item->widget()->maximumSize(); // Get the item's maximum allowable size.
+        // Constrain the item's width and height to its size hint or maximum size.
+        int itemWidth = qMin(item->sizeHint().width(), itemMaxSize.width());
+        int itemHeight = qMin(item->sizeHint().height(), itemMaxSize.height());
+        // Set the item's geometry based on the calculated size and position.
+        item->setGeometry(QRect(QPoint(x, y), QSize(itemWidth, itemHeight)));
+        // Move the x-position for the next item, including horizontal spacing.
+        x += itemWidth + horizontalSpacing();
+    }
+}
+
+/**
+ * @brief Returns the preferred size for this layout.
+ * @return The maximum of all item size hints as a QSize.
+ */
+QSize FlowLayout::sizeHint() const
+{
+    QSize size;
+    for (const QLayoutItem *item : items) {
+        if (!(item == nullptr || item->isEmpty())) {
+            size = size.expandedTo(item->sizeHint());
+        }
+    }
+    return size.isValid() ? size : QSize(0, 0);
+}
+
+/**
+ * @brief Returns the minimum size required to display all layout items.
+ * @return The minimum QSize needed by the layout.
+ */
+QSize FlowLayout::minimumSize() const
+{
+    QSize size;
+    for (const QLayoutItem *item : items) {
+        if (!(item == nullptr || item->isEmpty())) {
+            size = size.expandedTo(item->minimumSize());
+        }
+    }
+
+    size.setWidth(qMin(size.width(), getParentScrollAreaWidth()));
+    size.setHeight(qMin(size.height(), getParentScrollAreaHeight()));
+
+    return size.isValid() ? size : QSize(0, 0);
+}
+
+/**
+ * @brief Adds a new item to the layout.
+ * @param item The layout item to add.
+ */
+void FlowLayout::addItem(QLayoutItem *item)
+{
+    if (item != nullptr) {
+        items.append(item);
+    }
+}
+
+/**
+ * @brief Retrieves the count of items in the layout.
+ * @return The number of layout items.
+ */
+int FlowLayout::count() const
+{
+    // Ensure safe conversion by casting the maximum limit to the same type as items.size()
+    assert(items.size() <= static_cast<qsizetype>(std::numeric_limits<int>::max()));
```

Comment:
assertions outside of tests are dangerous and could halt the program. I'm not sure about this one.